bionic/core/task_execution.py

```diff
@@ -20,15 +20,43 @@ class TaskRunnerEntry:
     `TaskState` execution and tracking.
     """
 
-    def __init__(self, state, is_requested, dep_entries):
+    def __init__(self, state, dep_entries):
         self.stage = EntryStage.NEW
         self.dep_entries = dep_entries
 
         self.state = state
         self.future = None
-        self.is_requested = is_requested
         self.results_by_dnode = None
 
+        # Set by mark_requested().
+        # Includes self as an ancestor.
+        self.is_requested = False
+        self.has_unserializable_requested_ancestor = False
+
+    @property
+    def unfinished_deps(self):
+        return [
+            dep_entry
+            for dep_entry in self.dep_entries
+            if dep_entry.stage != EntryStage.COMPLETED
+        ]
+
+    def mark_requested(self):
+        self.is_requested = True
+
+        assert self.state.should_persist is not None
+        # If the entity in unserialized, mark all the descendants as having
+        # an unserializable requested ancestor.
+        if self.state.should_persist is False:
+            descendants = [self]
+            while len(descendants) > 0:
+                descendant = descendants.pop(0)
+                # Continue if the descendant was already marked.
+                if descendant.has_unserializable_requested_ancestor:
+                    continue
+                descendant.has_unserializable_requested_ancestor = True
+                descendants.extend(descendant.dep_entries)
```

Comment:
Wait -- aren't `dep_entries` the parents, not the children? It seems like we're recursing through the ancestors here, not the descendants.

Also, another things that concerns me: the normal way data flows in this code is by having each node pull down data from its parents in order to configure itself (immutably). Here a node is pushing data down to its descendants by writing mutable state. I don't understand why we need to reverse the direction in this case, and I don't understand why we're adding a mutable `is_requested` field that can return different values at different times.

Am I misunderstanding what's going on here?

Reply:
I'm not really happy with `has_unserializable_requested_ancestor` either. I added this because I needed some logic in place to make stack recursion work. I think this should simplify itself as we tackle tech debt around different states of a `TaskState`. Let's discuss this as well in our check-in.

Also, good point about ancestors and descendants. I was looking it from how execution stack uses them rather than how deps work.